colossalai/tensor/shape_consistency.py

```diff
@@ -0,0 +1,306 @@
+import torch
+from colossalai.tensor.sharding_spec import ShardingSpec, _DimSpec
+from enum import Enum
+from copy import deepcopy
+
+
+class CollectiveCommPattern(Enum):
+    ALLGATHER = 'all_gather'
+    ALLTOALL = 'all_to_all'
+    SHARD = 'shard'
+
+
+class ShapeConsistencyManager:
+
+    def __init__(self, consistency_option=None):
+        self.consistency_option = consistency_option
+        self.total_communication_cost = 0
+        self.total_transform_steps = 0
+        self.cached_spec_pairs = {}
+
+    def _all_gather_simulator(self, target_pair):
+        '''
+        Simulating all-gather operation, analyze the communication cost
+        and simulate the influence of the DimSpec.
+
+        We don't allow uncontiguous layout, such as all-gather(S012)->S02 is NOT allowed.
+        Therefore, all gather operation just remove the last element in shard list,
+        e.g.: 
+            all-gather(S01) -> S0
+        '''
+        pos, shard_list = target_pair
+        new_shard_list = shard_list[:-1]
+        # TODO: compute comm cost
+        comm_cost = 0
+        return (pos, new_shard_list), comm_cost
```

Comment:
`pos` is obtained from input and directly returned.

Reply:
I see. I will delete it.